tokio/src/fs/file.rs

```diff
@@ -756,3 +760,78 @@ impl Inner {
         }
     }
 }
+
+#[cfg(all(target_os = "linux", not(test)))]
+mod read_nowait {
+    use crate::io::ReadBuf;
+    use libc::{c_int, c_void, iovec, off_t, preadv2};
+    use std::{
+        os::unix::prelude::AsRawFd,
+        sync::atomic::{AtomicBool, Ordering},
+    };
+
+    static NONBLOCKING_READ_SUPPORTED: AtomicBool = AtomicBool::new(true);
+
+    pub(crate) fn try_nonblocking_read(
+        file: &crate::fs::sys::File,
+        dst: &mut ReadBuf<'_>,
+    ) -> Option<std::io::Result<()>> {
+        if !NONBLOCKING_READ_SUPPORTED.load(Ordering::Relaxed) {
+            return None;
+        }
+        let out = preadv2_safe(file, dst, -1, libc::RWF_NOWAIT);
+        if let Err(err) = &out {
+            match err.raw_os_error() {
+                Some(libc::ENOSYS) => {
+                    NONBLOCKING_READ_SUPPORTED.store(false, Ordering::Relaxed);
+                    return None;
+                }
+                Some(libc::ENOTSUP) | Some(libc::EAGAIN) => return None,
+                _ => {}
+            }
+        }
+        Some(out)
+    }
+
+    fn preadv2_safe(
+        file: &crate::fs::sys::File,
+        dst: &mut ReadBuf<'_>,
+        offset: off_t,
+        flags: c_int,
+    ) -> std::io::Result<()> {
+        unsafe {
+            /* We're manually have to defend against buffer overflows here.  The slice API makes
```

Comment:
Small fix: "We manually have to defend against buffer overflows here. The slice API makes this fairly straightforward."

Reply:
Thanks, fixed. straightforward is a real blind-spot for me ;)